interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+// and it would be necessary to make a deep comparison of two variants to tell
+// that they are the same. Since we want to map from variant to rule ID, we need
+// to use the variant string as the key.
+type variantEntry struct {
+	Variant patterns.PatternVariant
+	RuleID  prompting.IDType
+}
+
+// permissionDB stores a map from path pattern variant to the ID of the rule
+// associated with the variant for the permission associated with the permission
+// DB.
+type permissionDB struct {
+	// permissionDB contains a map from path pattern variant to rule ID
+	VariantEntries map[string]variantEntry
+}
+
+// interfaceDB stores a map from permission to the DB of rules pertaining to that
+// permission for the interface associated with the interface DB.
+type interfaceDB struct {
+	// interfaceDB contains a map from permission to permissionDB for a particular interface
+	PerPermission map[string]*permissionDB
+}
+
+// snapDB stores a map from interface name to the DB of rules pertaining to that
+// interface for the snap associated with the snap DB.
+type snapDB struct {
+	// snapDB contains a map from interface to interfaceDB for a particular snap
+	PerInterface map[string]*interfaceDB
+}
+
+// userDB stores a map from snap name to the DB of rules pertaining to that
+// snap for the user associated with the user DB.
+type userDB struct {
+	// userDB contains a map from snap to snapDB for a particular user
+	PerSnap map[string]*snapDB
+}
+
+// RuleDB stores a mapping from rule ID to rule, and a tree of rule IDs
+// searchable by user, snap, interface, permission, and pattern variant.
+type RuleDB struct {
+	mutex     sync.Mutex
+	maxIDMmap prompting.MaxIDMmap
+	ids       map[prompting.IDType]int
+	rules     []*Rule
+	perUser   map[uint32]*userDB
+	// notifyRule is a closure which will be called to record a notice when a
+	// rule is added, patched, or removed.
+	notifyRule func(userID uint32, ruleID prompting.IDType, data map[string]string) error
+}
+
+// New creates a new rule database, loads existing rules from the database file,
+// and returns the populated database.
+//
+// The given notifyRule closure will be called when a rule is added, modified,
+// expired, or removed. In order to guarantee the order of notices, notifyRule
+// is called with the prompt DB lock held, so it should not block for a
+// substantial amount of time (such as to lock and modify snapd state).
+func New(notifyRule func(userID uint32, ruleID prompting.IDType, data map[string]string) error) (*RuleDB, error) {
+	maxIDFilepath := filepath.Join(dirs.SnapdStateDir(dirs.GlobalRootDir), "request-rule-max-id")
+	maxIDMmap, err := prompting.OpenMaxIDMmap(maxIDFilepath)
+	if err != nil {
+		return nil, err
+	}
+	rdb := &RuleDB{
+		maxIDMmap:  maxIDMmap,
+		ids:        make(map[prompting.IDType]int),
+		rules:      make([]*Rule, 0),
+		perUser:    make(map[uint32]*userDB),
+		notifyRule: notifyRule,
+	}
+	return rdb, rdb.load()
```

Comment:
If the load fails, we'll start with a fresh rule DB, the error was meant to tell the caller that this occurred, but they're still supposed to use the rule DB. But this is of course incorrect, especially now that the mmap could result in a real error. I'll switch it to log an error instead of returning it if `load` fails.